Fix sparse compressed tensor invariants checks when nnz==0

pearu opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/115755

This PR is a step toward deprecating `torch.empty(..., layout=<sparse compressed tensor layout>)` that usage should be minimized as it will produce invalid tensors, see also https://github.com/pytorch/pytorch/issues/90695 .

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115826



cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer @jcaip

pearu committed:
Fix sparse compressed tensor invariants checks when nnz==0

[ghstack-poisoned]

pearu committed:
Update on "Fix sparse compressed tensor invariants checks when nnz==0"

Fixes https://github.com/pytorch/pytorch/issues/115755

This PR is a step toward eliminating `torch.empty(..., layout=<sparse compressed tensor layout>)` that usage should be minimized as it will produce invalid tensors.




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

pearu committed:
Update on "Fix sparse compressed tensor invariants checks when nnz==0"

Fixes https://github.com/pytorch/pytorch/issues/115755

This PR is a step toward deprecating `torch.empty(..., layout=<sparse compressed tensor layout>)` that usage should be minimized as it will produce invalid tensors, see also https://github.com/pytorch/pytorch/issues/90695 .




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -1277,14 +1278,51 @@ Tensor triu_indices_cpu(
     c10::optional<Device> device,
     c10::optional<bool> pin_memory) {
+  Layout layout_ = layout.value_or(Layout::Strided);
+  if (at::sparse_csr::is_sparse_compressed(layout_)) {
```
I'm not sure we can land this due to overhead. Can you use the dispatcher for this to register a new function?

pearu replied:
No. `torch.zeros` is factory function that takes no tensor arguments, hence the dispatcher mechanism is not applicable.

@cpuhrsch what kind of overhead are you referring here? `is_sparse_compressed` is a rather simple inline function containing a simple switch statement.
One possible optimization would be:
```c++
if (layout_ != Layout::Strided && at::sparse_csr::is_sparse_compressed(layout_)) {
  ...
}
... // original code
```
but I am not sure that there will be a considerable overhead in either cases.

amjames replied:
Yeah there should be no concerns with additional overhead from these change.

cpuhrsch reviewed (approved):
Great! Do we also check that values.numel() == 0 for the zeros constructor? I think it's implied by the tests, but just double checking to make sure.

pearu commented:
> Great! Do we also check that values.numel() == 0 for the zeros constructor? I think it's implied by the tests, but just double checking to make sure.

`values` shape is `plain_indices_and_values_size` that has the last dimension equal to `nnz` (`=0`). Since `numel` is computed as product of dimensions, the result of `values.numel()` is always `0` from the `zeros` constructor.

On the other hand, sparse tensor invariants checks (that are always executed when running tests) ensure that `values.shape[-1]` is equal to `crow_indices[-1]` and the `zeros` constructor calls `crow_indices.zero_()`. So, the answer is affirmative.

pearu commented:
@pytorchbot merge

pearu committed:
Update on "Fix sparse compressed tensor invariants checks when nnz==0"

Fixes https://github.com/pytorch/pytorch/issues/115755

This PR is a step toward deprecating `torch.empty(..., layout=<sparse compressed tensor layout>)` that usage should be minimized as it will produce invalid tensors, see also https://github.com/pytorch/pytorch/issues/90695 .




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

pearu committed:
Update on "Fix sparse compressed tensor invariants checks when nnz==0"

Fixes https://github.com/pytorch/pytorch/issues/115755

This PR is a step toward deprecating `torch.empty(..., layout=<sparse compressed tensor layout>)` that usage should be minimized as it will produce invalid tensors, see also https://github.com/pytorch/pytorch/issues/90695 .




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

pearu committed:
Update on "Fix sparse compressed tensor invariants checks when nnz==0"

Fixes https://github.com/pytorch/pytorch/issues/115755

This PR is a step toward deprecating `torch.empty(..., layout=<sparse compressed tensor layout>)` that usage should be minimized as it will produce invalid tensors, see also https://github.com/pytorch/pytorch/issues/90695 .




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

pearu committed:
Update on "Fix sparse compressed tensor invariants checks when nnz==0"

Fixes https://github.com/pytorch/pytorch/issues/115755

This PR is a step toward deprecating `torch.empty(..., layout=<sparse compressed tensor layout>)` that usage should be minimized as it will produce invalid tensors, see also https://github.com/pytorch/pytorch/issues/90695 .




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

pearu committed:
Update on "Fix sparse compressed tensor invariants checks when nnz==0"

Fixes https://github.com/pytorch/pytorch/issues/115755

This PR is a step toward deprecating `torch.empty(..., layout=<sparse compressed tensor layout>)` that usage should be minimized as it will produce invalid tensors, see also https://github.com/pytorch/pytorch/issues/90695 .




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]
